projects/hathor-wallet-headless/0004-tx-transaction.md

```diff
@@ -0,0 +1,493 @@
+
+For this, we can segregate the mentioned system in three areas:
+- User land
+- Interpreter
+- Lib land
+
+On *user land* we provides:
+- Instruction template interfaces
+- Instruction classes
+- Transaction template builder
+- Transaction template client
+
+On *interpreter* side we provides:
+- Wallet facade interpreter
+- Wallet service interpreter
+
+On *lib land* we provides:
+- All the existing methods for transaction preparation, creation and send transaction, including helpers and utils to create transaction components
+- New methods to assist the interpreters, if need
+
+To limit the scope of this work, we are going to focus on transactions of version `0x01` which means that we will not create tokens or NFTs.
+
+## User land
+
+### Instruction template interfaces
+[instruction-template-interfaces]: #instruction-template-interfaces
+
+**Basic:**
+- [BaseTemplateInstruction](#basetemplateinstruction): determine a instruction template type
+- InputOutputTemplateInstruction: can control an insertion position into the array
+- ActionTemplateInstruction: can determine over which component to act
+
+These basic template instructions are for *internal* use and define **required fields** to declare a *template instruction literal* object by the user.
+
+The following interfaces are describing each instruction type by its fields and possible values. A user can define a literal object of these types to design a final transaction.
+
+**Input:**
+- [UtxoInputTemplateInstruction](#utxoinputtemplateinstruction): can control any param to query a UTXO input
+- RawInputTemplateInstruction: can determine a known UTXO at the risk to be already spent
+
+**Output:**
+- [TokenOutputTemplateInstruction](#utxoinputtemplateinstruction): can control a token output generation
+- DataOutputTemplateInstruction: can control data output generation
+- RawOutputTemplateInstruction: can determine an output
+
+**Action:**
+- [ShuffleActionTemplateInstruction](#shuffleactiontemplateinstruction): can determine a shuffle over a target array, either an input or output
+- FillChangeActionTemplateInstruction: can determine a change output generation
+
+#### `BaseTemplateInstruction`
+[basetemplateinstruction]: #basetemplateinstruction
+
+```ts
+interface BaseTemplateInstruction {
+  readonly type: AllInstructionTypes|string;
+}
+```
+
+- `type`: select a type from all possible types among inputs, outputs and actions
+
+#### `InputOutputTemplateInstruction`
+
+```ts
+interface InputOutputTemplateInstruction {
+  /**
+   * A positive integer
+   * It allows -1 to denote the last position in the array
+   */
+  position: number;
+}
+```
+
+- `position`: which index to insert the input, default to `-1`
+  - [NOTE-1]: can be `-1` to insert at the end
+  - [NOTE-2]: the request will fail if position doesn't exists
```

Comment:
Otherwise we would request the position as required. We can set any position as default, this is an opinionated guess. Also, this default doesn't guarantee a final position as the user can define an action to shuffle it.

Reply:
Will remove the required comment for optional fields

Reply:
Done